File: backend/src/middleware/ingest-auth.ts
import type { FastifyReply, FastifyRequest } from "fastify";

export function requireIngestAuth(expectedToken: string) {
  return async function ingestAuth(
    request: FastifyRequest,
    reply: FastifyReply,
  ): Promise<void> {
    if (!expectedToken) {
      await reply.code(503).send({
        ok: false,
        message: "Local console ingest token is not configured.",
      });
      return;
    }

    const authorization = request.headers.authorization ?? "";
    if (authorization !== `Bearer ${expectedToken}`) {
      await reply.code(401).send({
        ok: false,
        message: "Unauthorized",
      });
    }
  };
}

File: backend/internal/middleware/ingestauth.go
package middleware

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// RequireIngestAuth mirrors backend/src/middleware/ingest-auth.ts.
func RequireIngestAuth(expectedToken string) gin.HandlerFunc {
	return func(c *gin.Context) {
		if expectedToken == "" {
			c.AbortWithStatusJSON(http.StatusServiceUnavailable, gin.H{
				"ok":      false,
				"message": "Local console ingest token is not configured.",
			})
			return
		}

		if c.GetHeader("Authorization") != "Bearer "+expectedToken {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"ok":      false,
				"message": "Unauthorized",
			})
			return
		}
		c.Next()
	}
}
